### PR TITLE
Highlight individual tokens in the source file and the corresponding nodes in the LLVM AST pane

### PR DESCRIPTION
**Pull request by necto**
Enable highlighting of spans of tokens in the source editor when the corresponding AST node is hovered over.
Enable the symmetric highlighting of the AST nodes corresponding to precisely the token being hovered over in the source editor.
In both cases `linked-code-decoration-column` marks the highlighted regions (in addition to `linked-code-decoration-margin` that works on the LOC granularity level).
Currently the `linked-code-decoration-column` style is "bold", so on the line highlighted with blue background, some tokens also become bold indicating the precise tokens corresponding to the AST node.

Incidentally, this also fixes the display of the span corresponding to a selected warning/error in the compiler output pane. Up until this change, only the first character has been highlighted when a user hovers the pointer over the warning. Now, the entire token is highlighted.

### CODE REVIEW

**Comment by necto** [comment]
Pay attention to the bolt font, it indicates the precise highlighting (in addition to the blue background indicating the coarse-grained highlighting)
llvm-ast-linked-lines-highlight

**Comment by necto** [comment]
Some of the decisions here might be controversial, I'd love to see your input on:
* bolding the entire token rather than just one character when showing an error. See the end of the gif above.
* computing the end for squiggly-underlined span manually (using `getTokenSpan`) instead of the Monaco built-in (triggered by  -1 passed as the end column). In particular, stretch underlining to the end of the string literal rather than just one character. See the screenshots:
before:
Screenshot from 2021-04-04 12-46-57
after:
Screenshot from 2021-04-04 12-38-31

**Review by RubenRBS — APPROVED** [review]
Omg I love this, thank you very much!

As for the bloding and squiggly changes, I also love them.

File: static/panes/ir-view.js [hunk 1]
```diff
@@ -260,8 +260,8 @@ Ir.prototype.onMouseMove = function (e) {
         if (hoverCode) {
             // We check that we actually have something to show at this point!
             var sourceLine = hoverCode.source && !hoverCode.source.file ? hoverCode.source.line : -1;
-            this.eventHub.emit('editorLinkLine', this._editorid, sourceLine, -1, false);
-            this.eventHub.emit('panesLinkLine', this._compilerid, sourceLine, false, this.getPaneName());
+            this.eventHub.emit('editorLinkLine', this._editorid, sourceLine, -1, -1, false);
+            this.eventHub.emit('panesLinkLine', this._compilerid, sourceLine, -1, -1, false, this.getPaneName());
```
Comment (mattgodbolt) [inline static/panes/ir-view.js:264]: Thanks for finding and updating these too!

File: test/llvm-ast-parser-tests.js [hunk 2]
```diff
@@ -80,9 +80,15 @@ describe('llvm-ast', function () {
         should.exist(compilerOutput.stdout.find(l => l.text.match(/col:21, line:4:1/)));
         should.exist(compilerOutput.stdout.find(l => l.text.match(/line:3:5, col:18/)));
         let processed = astParser.processAst(cloneDeep(compilerOutput));
-        should.exist(processed.find(l => l.source && 0 < l.source.from));
-        processed.find(l => l.text.match(/col:21, line:4:1/)).source.to.should.equal(4);
-        processed.find(l => l.text.match(/line:3:5, col:18/)).source.from.should.equal(3);
-        processed.find(l => l.text.match(/line:3:5, col:18/)).source.to.should.equal(3);
+        should.exist(processed.find(l => l.source && 0 < l.source.from.line));
```
Comment (mattgodbolt) [inline test/llvm-ast-parser-tests.js:83]: Thanks for the tests!

File: static/panes/editor.js [hunk 3]
```diff
@@ -1029,7 +1041,46 @@ Editor.prototype.onSelectLine = function (id, lineNum) {
...
+// that spans across the column.
+// a - colStart points to the first character of the token
+// b - colEnd points to the character immediately following the token
+// e.g.: "this->callableMethod ( x, y );"
+//              ^a   ^column  ^b
+Editor.prototype.getTokenSpan = function (lineNum, column) {
+    var model = this.editor.getModel();
+    var line = model.getLineContent(lineNum);
+    if (0 < column && column < line.length) {
+        var tokens = monaco.editor.tokenize(line, model.getModeId());
```
Comment (mattgodbolt) [inline static/panes/editor.js:1054]: It's so unfortunate we have to do this. I have another project which needed the same sad piece of code. Thanks for taking the time to work out how this is done!

File: static/panes/editor.js [hunk 4]
```diff
@@ -1029,7 +1041,46 @@ Editor.prototype.onSelectLine = function (id, lineNum) {
     }
 };
 
-Editor.prototype.onEditorLinkLine = function (editorId, lineNum, columnNum, reveal) {
+// Returns a half-segment [a, b) for the token on the line lineNum
```
Comment (mattgodbolt) [inline static/panes/editor.js:1044]: Thanks for the elucidating comment! This stuff is super tricksy

**Review by mattgodbolt — APPROVED** [review]
This is fantastic! Thanks so much!

**Comment by mattgodbolt** [comment]
This is now live! Thank you!

**Comment by necto** [comment]
Awesome! Thank you guys for Compiler Explorer!

Matt Godbolt ***@***.***> schrieb am Di., 6. Apr. 2021, 22:59:

**Pull request merged**